src/3d/qgs3dsceneexporter.cpp

```diff
@@ -0,0 +1,524 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dsceneexporter.h"
+
+#include <QVector>
+#include <Qt3DCore/QEntity>
+#include <Qt3DCore/QComponent>
+#include <Qt3DCore/QNode>
+#include <Qt3DRender/QGeometry>
+#include <Qt3DRender/QAttribute>
+#include <Qt3DRender/QBuffer>
+#include <Qt3DRender/QGeometryRenderer>
+#include <Qt3DExtras/QPlaneGeometry>
+#include <Qt3DCore/QTransform>
+#include <Qt3DRender/QMaterial>
+#include <Qt3DExtras/QDiffuseMapMaterial>
+#include <Qt3DExtras/QTextureMaterial>
+#include <Qt3DRender/QTextureImage>
+#include <Qt3DRender/QTexture>
+
+#include <QByteArray>
+#include <QFile>
+#include <QTextStream>
+
+#include "qgstessellatedpolygongeometry.h"
+#include "qgsterraintileentity_p.h"
+#include "qgsterrainentity_p.h"
+#include "qgschunknode_p.h"
+#include "qgsterraingenerator.h"
+#include "qgs3dmapsettings.h"
+#include "qgsflatterraingenerator.h"
+#include "qgsdemterraingenerator.h"
+#include "qgsdemterraintileloader_p.h"
+#include "qgsdemterraintilegeometry_p.h"
+#include "qgs3dexportobject.h"
+#include "qgsterraintextureimage_p.h"
+#include "qgsterraintexturegenerator_p.h"
+
+#include <numeric>
+
+template<typename T>
+QVector<T> getAttributeData( Qt3DRender::QAttribute *attribute )
+{
+  QByteArray data = attribute->buffer()->data();
+  uint bytesOffset = attribute->byteOffset();
+  uint bytesStride = attribute->byteStride();
+  uint vertexSize = attribute->vertexSize();
+
+  QVector<T> result;
+  for ( int i = bytesOffset; i < data.size(); i += bytesStride )
+  {
+    for ( unsigned int j = 0; j < vertexSize * sizeof( T ); j += sizeof( T ) )
+    {
+      // maybe a problem with indienness can happen?
+      T v;
+      char *vArr = ( char * )&v;
+      for ( unsigned int k = 0; k < sizeof( T ); ++k )
+      {
+        vArr[k] = data.at( i + j + k );
+      }
+      result.push_back( v );
+    }
+  }
+  return result;
+}
+
+template<>
+QVector<unsigned int> getAttributeData<unsigned int>( Qt3DRender::QAttribute *attribute )
+{
+  QByteArray data = attribute->buffer()->data();
+
+  QVector<unsigned int> result;
+  for ( int i = 0; i < data.size(); i += sizeof( unsigned int ) )
+  {
+    // maybe a problem with indienness can happen?
+    unsigned int v;
+    char *vArr = ( char * )&v;
+    for ( int k = 0; k < ( int )sizeof( unsigned int ); ++k )
+    {
+      vArr[k] = data.at( i + k );
+    }
+    result.push_back( v );
+  }
+
+  return result;
+}
+
+QVector<float> createPlaneVertexData( float w, float h, const QSize &resolution )
```

Comment:
do we need these createPlane*Data() methods? Can't we just extract the contents of attributes of the plane geometry created for terrain tiles?